// @inlang/sdk/dist/createMessagesQuery.js
import { ReactiveMap } from "./reactivity/map.js";
import { createEffect } from "./reactivity/solid.js";
import { createSubscribable } from "./loadProject.js";
/**
 * Creates a reactive query API for messages.
 */
export function createMessagesQuery(messages) {
    // @ts-expect-error
    const index = new ReactiveMap();
    // Map default alias to message
    // Assumes that aliases are only created and deleted, not updated
    // TODO #2346 - handle updates to aliases
    // TODO #2346 - refine to hold messageId[], if default alias is not unique
    // @ts-expect-error
    const defaultAliasIndex = new ReactiveMap();
    createEffect(() => {
        index.clear();
        for (const message of structuredClone(messages())) {
            index.set(message.id, message);
            if ("default" in message.alias) {
                defaultAliasIndex.set(message.alias.default, message);
            }
        }
    });
    const get = (args) => index.get(args.where.id);
    const getByDefaultAlias = (alias) => defaultAliasIndex.get(alias);
    return {
        create: ({ data }) => {
            if (index.has(data.id))
                return false;
            index.set(data.id, data);
            if ("default" in data.alias) {
                defaultAliasIndex.set(data.alias.default, data);
            }
            return true;
        },
        get: Object.assign(get, {
            subscribe: (args, callback) => createSubscribable(() => get(args)).subscribe(callback),
        }),
        getByDefaultAlias: Object.assign(getByDefaultAlias, {
            subscribe: (alias, callback) => createSubscribable(() => getByDefaultAlias(alias)).subscribe(callback),
        }),
        includedMessageIds: createSubscribable(() => {
            return [...index.keys()];
        }),
        getAll: createSubscribable(() => {
            return [...index.values()];
        }),
        update: ({ where, data }) => {
            const message = index.get(where.id);
            if (message === undefined)
                return false;
            index.set(where.id, { ...message, ...data });
            return true;
        },
        upsert: ({ where, data }) => {
            const message = index.get(where.id);
            if (message === undefined) {
                index.set(where.id, data);
                if ("default" in data.alias) {
                    defaultAliasIndex.set(data.alias.default, data);
                }
            }
            else {
                index.set(where.id, { ...message, ...data });
            }
            return true;
        },
        delete: ({ where }) => {
            const message = index.get(where.id);
            if (message === undefined)
                return false;
            if ("default" in message.alias) {
                defaultAliasIndex.delete(message.alias.default);
            }
            index.delete(where.id);
            return true;
        },
    };
}
