// @inlang/sdk/dist/createNodeishFsWithWatcher.js
/**
 * Wraps the nodeish filesystem subset with a function that intercepts paths
 * and prepends the base path.
 *
 * The paths are resolved from the `projectPath` argument.
 */
export const createNodeishFsWithWatcher = (args) => {
    const pathList = [];
    const makeWatcher = (path) => {
        const abortController = new AbortController();
        (async () => {
            try {
                const watcher = args.nodeishFs.watch(path, {
                    signal: abortController.signal,
                    persistent: false,
                });
                if (watcher) {
                    //eslint-disable-next-line @typescript-eslint/no-unused-vars
                    for await (const event of watcher) {
                        args.updateMessages();
                    }
                }
            }
            catch (err) {
                if (err.name === "AbortError")
                    return;
                // https://github.com/inlang/monorepo/issues/1647
                // the file does not exist (yet)
                // this is not testable beacause the fs.watch api differs
                // from node and lix. lenghty
                else if (err.code === "ENOENT")
                    return;
                throw err;
            }
        })();
    };
    const readFileAndExtractPath = (path, options) => {
        if (!pathList.includes(path)) {
            makeWatcher(path);
            pathList.push(path);
        }
        return args.nodeishFs.readFile(path, options);
    };
    return {
        // @ts-expect-error
        readFile: (path, options) => readFileAndExtractPath(path, options),
        readdir: args.nodeishFs.readdir,
        mkdir: args.nodeishFs.mkdir,
        writeFile: args.nodeishFs.writeFile,
        watch: args.nodeishFs.watch,
    };
};
