// @inlang/sdk/dist/createMessageLintReportsQuery.js
import { createEffect } from "./reactivity/solid.js";
import { createSubscribable } from "./loadProject.js";
import { lintSingleMessage } from "./lint/index.js";
import { ReactiveMap } from "./reactivity/map.js";
/**
 * Creates a reactive query API for messages.
 */
export function createMessageLintReportsQuery(messages, settings, installedMessageLintRules, resolvedModules) {
    // @ts-expect-error
    const index = new ReactiveMap();
    createEffect(() => {
        const modules = resolvedModules();
        const _messages = messages();
        const _settings = settings();
        if (_messages && _settings && modules) {
            // index.clear()
            for (const message of _messages) {
                // TODO: only lint changed messages and update arrays selectively
                lintSingleMessage({
                    rules: modules.messageLintRules,
                    settings: {
                        ..._settings,
                        messageLintRuleLevels: Object.fromEntries(installedMessageLintRules().map((rule) => [rule.id, rule.level])),
                    },
                    messages: _messages,
                    message: message,
                }).then((report) => {
                    if (report.errors.length === 0 &&
                        JSON.stringify(index.get(message.id)) !== JSON.stringify(report.data)) {
                        index.set(message.id, report.data || []);
                    }
                });
            }
        }
    });
    const get = (args) => {
        return structuredClone(index.get(args.where.messageId));
    };
    return {
        getAll: createSubscribable(() => {
            return structuredClone([...index.values()].flat().length === 0 ? [] : [...index.values()].flat());
        }),
        get: Object.assign(get, {
            subscribe: (args, callback) => createSubscribable(() => get(args)).subscribe(callback),
        }),
    };
}
