// @inlang/sdk/dist/createMessagesQuery.js
import { ReactiveMap } from "./reactivity/map.js";
import { createEffect } from "./reactivity/solid.js";
import { createSubscribable } from "./loadProject.js";
/**
 * Creates a reactive query API for messages.
 */
export function createMessagesQuery(messages) {
    // @ts-expect-error
    const index = new ReactiveMap();
    createEffect(() => {
        index.clear();
        for (const message of structuredClone(messages())) {
            index.set(message.id, message);
        }
    });
    const get = (args) => index.get(args.where.id);
    return {
        create: ({ data }) => {
            if (index.has(data.id))
                return false;
            index.set(data.id, data);
            return true;
        },
        get: Object.assign(get, {
            subscribe: (args, callback) => createSubscribable(() => get(args)).subscribe(callback),
        }),
        includedMessageIds: createSubscribable(() => {
            return [...index.keys()];
        }),
        getAll: createSubscribable(() => {
            return [...index.values()];
        }),
        update: ({ where, data }) => {
            const message = index.get(where.id);
            if (message === undefined)
                return false;
            index.set(where.id, { ...message, ...data });
            return true;
        },
        upsert: ({ where, data }) => {
            const message = index.get(where.id);
            if (message === undefined) {
                index.set(where.id, data);
            }
            else {
                index.set(where.id, { ...message, ...data });
            }
            return true;
        },
        delete: ({ where }) => {
            if (!index.has(where.id))
                return false;
            index.delete(where.id);
            return true;
        },
    };
}
