// @inlang/sdk/dist/resolve-modules/resolveModules.js
import { InlangModule } from "@inlang/module";
import { ModuleError, ModuleImportError, ModuleHasNoExportsError, ModuleExportIsInvalidError, ModuleSettingsAreInvalidError, } from "./errors.js";
import { tryCatch } from "@inlang/result";
import { resolveMessageLintRules } from "./message-lint-rules/resolveMessageLintRules.js";
import { createImport } from "./import.js";
import { resolvePlugins } from "./plugins/resolvePlugins.js";
import { TypeCompiler } from "@sinclair/typebox/compiler";
import { validatedModuleSettings } from "./validatedModuleSettings.js";
const ModuleCompiler = TypeCompiler.Compile(InlangModule);
export const resolveModules = async (args) => {
    const _import = args._import ?? createImport({ readFile: args.nodeishFs.readFile });
    const moduleErrors = [];
    const allPlugins = [];
    const allMessageLintRules = [];
    const meta = [];
    for (const module of args.settings.modules) {
        /**
         * -------------- BEGIN SETUP --------------
         */
        const importedModule = await tryCatch(() => _import(module));
        // -- IMPORT MODULE --
        if (importedModule.error) {
            moduleErrors.push(new ModuleImportError({
                module: module,
                cause: importedModule.error,
            }));
            continue;
        }
        // -- MODULE DOES NOT EXPORT ANYTHING --
        if (importedModule.data?.default === undefined) {
            moduleErrors.push(new ModuleHasNoExportsError({
                module: module,
            }));
            continue;
        }
        // -- CHECK IF MODULE IS SYNTACTIALLY VALID
        const isValidModule = ModuleCompiler.Check(importedModule.data);
        if (isValidModule === false) {
            const errors = [...ModuleCompiler.Errors(importedModule.data)];
            moduleErrors.push(new ModuleExportIsInvalidError({
                module: module,
                errors,
            }));
            continue;
        }
        // -- VALIDATE MODULE SETTINGS
        const result = validatedModuleSettings({
            settingsSchema: importedModule.data.default.settingsSchema,
            moduleSettings: args.settings[importedModule.data.default.id],
        });
        if (result !== "isValid") {
            moduleErrors.push(new ModuleSettingsAreInvalidError({ module: module, errors: result }));
            continue;
        }
        meta.push({
            module: module,
            id: importedModule.data.default.id,
        });
        if (importedModule.data.default.id.startsWith("plugin.")) {
            allPlugins.push(importedModule.data.default);
        }
        else if (importedModule.data.default.id.startsWith("messageLintRule.")) {
            allMessageLintRules.push(importedModule.data.default);
        }
        else {
            moduleErrors.push(new ModuleError(`Unimplemented module type ${importedModule.data.default.id}.The module has not been installed.`, { module: module }));
        }
    }
    const resolvedPlugins = await resolvePlugins({
        plugins: allPlugins,
        settings: args.settings,
        nodeishFs: args.nodeishFs,
    });
    const resolvedLintRules = resolveMessageLintRules({ messageLintRules: allMessageLintRules });
    return {
        meta,
        messageLintRules: allMessageLintRules,
        plugins: allPlugins,
        resolvedPluginApi: resolvedPlugins.data,
        errors: [...moduleErrors, ...resolvedLintRules.errors, ...resolvedPlugins.errors],
    };
};
